File: tools/clod-poc/src/world_source/world_source.ts
import * as THREE from "three";
import { surfaceHeightCore } from "../gpu/terrain_field_core.js";
import {
  DEFAULT_TERRAIN_FIELD_CONFIG,
  resolveTerrainFieldConfig,
  surfaceHeight,
  type TerrainFieldConfig,
  type TerrainFieldConfigInput,
} from "../terrain/terrain.js";
import { BiomeRegionField, type BiomeId } from "./biome_region_field.js";
import { sampleIslandMask } from "./island_shape.js";

export interface WorldSourceMetadata {
  seed: number;
  seaLevel: number;
  bounds: "infinite" | { radiusM: number };
  oceanRim: boolean;
  terrain: TerrainFieldConfig;
}

export interface WorldSource {
  readonly metadata: WorldSourceMetadata;
  sampleHeight(x: number, z: number): number;
  sampleBiome(x: number, z: number): BiomeId;
  sampleMaterial(x: number, z: number): number;
  oceanMask(x: number, z: number): number;
}

export interface FarWorldHeightProvider {
  sampleHeight(x: number, z: number): number;
  sampleNormal(x: number, z: number): THREE.Vector3;
}

function terrainDefaultBounds(terrain: TerrainFieldConfig): WorldSourceMetadata["bounds"] {
  return terrain.islandShape.oceanRim ? { radiusM: terrain.islandShape.worldRadiusM } : "infinite";
}

function resolveMetadataTerrain(metadata: Partial<WorldSourceMetadata>): TerrainFieldConfig {
  const base = metadata.terrain ?? DEFAULT_TERRAIN_FIELD_CONFIG;
  const seed = metadata.seed ?? base.seed;
  const seaLevel = metadata.seaLevel ?? base.seaLevel;
  const oceanRim = metadata.oceanRim ?? base.islandShape.oceanRim;
  return resolveTerrainFieldConfig({
    ...base,
    seed,
    seaLevel,
    islandShape: {
      ...base.islandShape,
      oceanRim,
      seed,
      seaLevel,
    },
  });
}

function notImplemented(method: string): never {
  throw new Error(`StreamedVoxelWorldSource.${method} is not implemented yet`);
}

export class ProceduralWorldSource implements WorldSource {
  readonly metadata: WorldSourceMetadata;
  private readonly biomes: BiomeRegionField;

  constructor(terrain: TerrainFieldConfigInput = DEFAULT_TERRAIN_FIELD_CONFIG) {
    const resolvedTerrain = resolveTerrainFieldConfig(terrain);
    this.metadata = {
      seed: resolvedTerrain.seed,
      seaLevel: resolvedTerrain.seaLevel,
      bounds: terrainDefaultBounds(resolvedTerrain),
      oceanRim: resolvedTerrain.islandShape.oceanRim,
      terrain: resolvedTerrain,
    };
    this.biomes = new BiomeRegionField({
      seed: resolvedTerrain.seed,
      seaLevel: resolvedTerrain.seaLevel,
      islandShape: resolvedTerrain.islandShape,
    });
  }

  sampleHeight(x: number, z: number): number {
    return surfaceHeightCore(x, z, this.metadata.terrain);
  }

  sampleBiome(x: number, z: number): BiomeId {
    return this.biomes.sample(x, z, this.sampleHeight(x, z)).biome;
  }

  sampleMaterial(x: number, z: number): number {
    return this.sampleBiome(x, z);
  }

  oceanMask(x: number, z: number): number {
    const h = this.sampleHeight(x, z);
    const island = sampleIslandMask(x, z, this.metadata.terrain.islandShape);
    if (h < this.metadata.seaLevel) return 1;
    return Math.max(0, Math.min(1, 1 - island.mask));
  }

  createFarHeightProvider(): FarWorldHeightProvider {
    return {
      sampleHeight: (x, z) => this.sampleHeight(x, z),
      sampleNormal: (x, z) => {
        const e = 2;
        const hL = this.sampleHeight(x - e, z);
        const hR = this.sampleHeight(x + e, z);
        const hD = this.sampleHeight(x, z - e);
        const hU = this.sampleHeight(x, z + e);
        return new THREE.Vector3((hL - hR) / (2 * e), 1, (hD - hU) / (2 * e)).normalize();
      },
    };
  }
}

/**
 * Runtime world authority. Unlike ProceduralWorldSource, this source follows the
 * installed heightfield/hydrology authority, so far terrain, biomes, materials,
 * props, collision queries, and near terrain classify the same surface.
 */
export class CanonicalWorldSource extends ProceduralWorldSource {
  override sampleHeight(x: number, z: number): number {
    return surfaceHeight(x, z);
  }
}

export class StreamedVoxelWorldSource implements WorldSource {
  readonly metadata: WorldSourceMetadata;

  constructor(metadata: Partial<WorldSourceMetadata> = {}) {
    const terrain = resolveMetadataTerrain(metadata);
    this.metadata = {
      seed: terrain.seed,
      seaLevel: terrain.seaLevel,
      bounds: metadata.bounds ?? terrainDefaultBounds(terrain),
      oceanRim: terrain.islandShape.oceanRim,
      terrain,
    };
  }

  sampleHeight(_x: number, _z: number): number {
    return notImplemented("sampleHeight");
  }

  sampleBiome(_x: number, _z: number): BiomeId {
    return notImplemented("sampleBiome");
  }

  sampleMaterial(_x: number, _z: number): number {
    return notImplemented("sampleMaterial");
  }

  oceanMask(_x: number, _z: number): number {
    return notImplemented("oceanMask");
  }
}

File: src/world/source/world_source.rs
use bevy::prelude::Resource;
use serde::{Deserialize, Serialize};

use super::biome_region_field::{BiomeId, BiomeRegionField};
use super::island_shape::{sample_island_mask, IslandShapeConfig};
use super::height_field::base_surface_height;

pub const DEFAULT_TERRAIN_SEED: i32 = 0;
pub const DEFAULT_SEA_LEVEL: f32 = 18.0;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TerrainFieldConfig {
    pub seed: i32,
    pub sea_level: f32,
    pub island_shape: IslandShapeConfig,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum WorldSourceBounds {
    Infinite,
    RadiusM(f32),
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct WorldSourceMetadata {
    pub seed: i32,
    pub sea_level: f32,
    pub bounds: WorldSourceBounds,
    pub ocean_rim: bool,
    pub terrain: TerrainFieldConfig,
}

pub trait WorldSource {
    fn metadata(&self) -> &WorldSourceMetadata;
    fn sample_height(&self, x: f32, z: f32) -> f32;
    fn sample_biome(&self, x: f32, z: f32) -> BiomeId;
    fn ocean_mask(&self, x: f32, z: f32) -> f32;
}

#[derive(Resource, Debug, Clone)]
pub struct ProceduralWorldSource {
    metadata: WorldSourceMetadata,
    biomes: BiomeRegionField,
}

impl TerrainFieldConfig {
    pub fn new(seed: i32, sea_level: f32, island_shape: IslandShapeConfig) -> Self {
        let mut island_shape = island_shape.sanitized();
        island_shape.sea_level = sea_level;
        Self { seed, sea_level, island_shape }
    }
}

impl Default for TerrainFieldConfig {
    fn default() -> Self {
        Self::new(DEFAULT_TERRAIN_SEED, DEFAULT_SEA_LEVEL, IslandShapeConfig::default())
    }
}

impl ProceduralWorldSource {
    pub fn new(terrain: TerrainFieldConfig) -> Self {
        let bounds = if terrain.island_shape.ocean_rim {
            WorldSourceBounds::RadiusM(terrain.island_shape.world_radius_m)
        } else {
            WorldSourceBounds::Infinite
        };
        let metadata = WorldSourceMetadata {
            seed: terrain.seed,
            sea_level: terrain.sea_level,
            bounds,
            ocean_rim: terrain.island_shape.ocean_rim,
            terrain: terrain.clone(),
        };
        let biomes = BiomeRegionField::new(terrain.seed, terrain.sea_level, terrain.island_shape.clone());
        Self { metadata, biomes }
    }

    pub fn biome_field(&self) -> &BiomeRegionField {
        &self.biomes
    }
}

impl Default for ProceduralWorldSource {
    fn default() -> Self {
        Self::new(TerrainFieldConfig::default())
    }
}

impl WorldSource for ProceduralWorldSource {
    fn metadata(&self) -> &WorldSourceMetadata {
        &self.metadata
    }

    fn sample_height(&self, x: f32, z: f32) -> f32 {
        base_surface_height(x, z, &self.metadata.terrain)
    }

    fn sample_biome(&self, x: f32, z: f32) -> BiomeId {
        self.biomes.sample(x, z, self.sample_height(x, z)).biome
    }

    fn ocean_mask(&self, x: f32, z: f32) -> f32 {
        let height = self.sample_height(x, z);
        let island = sample_island_mask(x, z, &self.metadata.terrain.island_shape);
        if height < self.metadata.sea_level {
            1.0
        } else {
            (1.0 - island.mask).clamp(0.0, 1.0)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn procedural_world_source_is_deterministic() {
        let source = ProceduralWorldSource::default();
        let a = source.sample_height(512.0, -128.0);
        let b = source.sample_height(512.0, -128.0);
        assert_eq!(a, b);
        assert_eq!(source.sample_biome(512.0, -128.0), source.sample_biome(512.0, -128.0));
    }

    #[test]
    fn ocean_rim_sets_bounded_metadata() {
        let source = ProceduralWorldSource::new(TerrainFieldConfig::new(
            0,
            18.0,
            IslandShapeConfig { ocean_rim: true, ..IslandShapeConfig::default() },
        ));
        assert!(matches!(source.metadata().bounds, WorldSourceBounds::RadiusM(_)));
        assert!(source.metadata().ocean_rim);
    }
}
